experimental/auth-logic/provenance_build_wrapper.go

```diff
@@ -0,0 +1,64 @@
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package authlogic contains logic and tests for interfacing with the
+// authorization logic compiler
+package authlogic
+
+import (
+	"fmt"
+
+	"github.com/project-oak/transparent-release/common"
+	"github.com/project-oak/transparent-release/slsa"
+)
+
+type provenanceBuildWrapper struct{ provenanceFilePath string }
+
+func (pbw provenanceBuildWrapper) EmitStatement() UnattributedStatement {
+	handleErr := func(err error) {
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	// Unmarshal a provenance struct from the JSON file
+	provenance, provenanceParseErr := slsa.ParseProvenanceFile(pbw.provenanceFilePath)
+	handleErr(provenanceParseErr)
+
+	applicationName := provenance.Subject[0].Name
+
+	// Generate a BuildConfig struct from the provenance file
+	buildConfig, loadBuildErr := common.LoadBuildConfigFromProvenance(provenance)
+	handleErr(loadBuildErr)
+
+	// Fetch the repository sources from the repository referenced in the
+	// BuildConfig struct.
+	_, repoFetchErr := common.FetchSourcesFromRepo(buildConfig.Repo, buildConfig.CommitHash)
+	handleErr(repoFetchErr)
+
+	// Build the binary from the fetched sources.
+	buildErr := buildConfig.Build()
+	handleErr(buildErr)
+
+	// Measure the hash of the binary.
+	measuredBinaryHash, hashErr := buildConfig.ComputeBinarySha256Hash()
+	handleErr(hashErr)
```

Comment:
To check my understanding: this is currently re-building the binary, because the provenance files are not signed, right? But eventually we want to rely on just inspecting the signature (by trusting some key to be associated with valid provenance files). I would expect the auth logic process (eventually) to not have to do any "actual" work, just check signatures and run authorization logic -- in particular, it need not know how to actually build a binary.

Reply:
Here's the mental model I'm building this with. There are three main parties:
 - the app developer / release publisher
 - the verifier
 - the client

The code being implemented here automates the verifier process. Part of the verifier process involves checking that the provenance file can really be used to build a binary and that the resulting binary has the expected hash. This wrapper is implementing that specific piece.

The top-level binary for this verification process using authorization logic will be both a boolean (which can be useful for example to write a bazel rule that just checks if some set of evidence passes the verification step) and another authorization logic credential that can be consumed by the client if desired. The credential will be a statement by some verifier principal that the hash is an acceptable one (it can also pick up the valid time ranges from endorsement files).